glyphs-reader/data/update.py

```diff
@@ -9,22 +9,75 @@
 """
 
 from collections import defaultdict
+from collections.abc import Callable
 import dataclasses
 from dataclasses import dataclass
 import glyphsLib
 from importlib import resources
 from lxml import etree
 from pathlib import Path
 from textwrap import dedent
-from typing import Iterable, Mapping, Optional, Tuple
+from typing import Iterable, Optional, Self, Tuple
 
 
 @dataclass(frozen=True)
 class GlyphInfo:
-    codepoint: Optional[str]
+    codepoint: Optional[int]
     name: str
     category: str
     subcategory: Optional[str]
+    production_name: Optional[str]
+    script: Optional[str]
+
+    def has_codepoint(self) -> bool:
+        return self.codepoint is not None
+    
+    def production_name_or_empty(self) -> str:
+        if self.production_name is None:
+            return ""
+        return self.production_name
+    
+    def production_name_file_bytes(self) -> bytes:
+        return ProdNamePattern.for_glyph_info(self).file_bytes(self)
+
+
+
+@dataclass(frozen=True)
+class ProdNamePattern:
+    display_name: str
+    pred: Callable[[GlyphInfo], bool]
+    file_bytes: Callable[[GlyphInfo], bytes]
+
+    def for_glyph_info(gi: GlyphInfo) -> Self:
+        for pattern in _PROD_NAME_PATTERNS:
+            if pattern.pred(gi):
+                return pattern
+        raise ValueError(f"At least 1 pattern should always match, somehow {gi} defeated us")
+
+# Production names are heavily patterned, try to take advantage to compress representation
+# Meant to be probed in order
+_PROD_NAME_PATTERNS = (
+    ProdNamePattern(
+        "blank",
+        lambda gi: gi.production_name is None,
+        lambda _gi: b"b",
+    ),
+    ProdNamePattern(
+        "uni-prefix",
+        lambda gi: gi.has_codepoint() and gi.production_name == f"uni{gi.codepoint:04X}",
+        lambda gi: b"u" + gi.codepoint.to_bytes(3, "little"),
+    ),
+    ProdNamePattern(
+        "u-prefix", 
+        lambda gi: gi.has_codepoint() and gi.production_name == f"u{gi.codepoint:04X}",
+        lambda gi: b"v" + gi.codepoint.to_bytes(3, "little"),
+    ),
```

Comment:
`uni` and `u`-prefixed names names can be generated at runtime from the codepoint, so may be check if the name would match the what would be generated and skip it entirely?

Reply:
Early hackery ... one day I'll find time to finish it :) We don't know the pattern from the codepoint so we at the absolute minimum needto know that this is a u-prefix style name vs uni-prefix vs some other random thing?

Reply:
`uni`-prefix is for BMP codepoints (i.e. <= 0xFFFF), `u`-prefix is for other codepoints.

Reply:
Oh ty, I hadn't realized they had that pattern. There are also 2490 that deviate from both patterns and 45425 that just don't have a prod name (`{'blank': 45425, 'custom': 2490, 'uni-prefix': 24776, 'u-prefix': 5529}` according to a quick and dirty count)

Reply:
I’m fairly sure (almost?) all production glyph names can be algorithmically generated. You basically need to the reverse of https://github.com/adobe-type-tools/agl-specification?tab=readme-ov-file#2-the-mapping, since this is how they are going to be consumed.

Reply:
I appear to be ending up with 2490 violations of the pattern based on ^. Saves a bunch of space over my original hack, ty for the ping!